File: 2020/julia/AOC/src/parsers.jl
module Parsers

const DATA = joinpath(@__DIR__, "..", "..", "..", "data")

day1() = parse.(Int, readlines(joinpath(DATA, "day1.txt")))

day2() = readlines(joinpath(DATA, "day2.txt"))

function day3()
	ret = map(readlines(joinpath(DATA, "day3.txt"))) do line
		[Int(x == '#') for x in strip(line)]
	end
	permutedims(reduce(hcat, ret))
end

function day4()
	items = String[]
	last_was_blank = true
	for line in eachline(joinpath(DATA, "day4.txt"), keep=true)
		if last_was_blank
			push!(items, line)
			last_was_blank = false
		elseif strip(line) == ""
			last_was_blank = true
		else
			items[end] *= line 
			last_was_blank = false
		end
	end
	items
end

day5() = eachline(joinpath(DATA, "day5.txt"))

function day6()
	items = Vector{Vector{String}}()
	last_was_blank = true
	for line in eachline(joinpath(DATA, "day6.txt"), keep=true)
		if last_was_blank
			push!(items, [strip(line)])
			last_was_blank = false
		elseif strip(line) == ""
			last_was_blank = true
		else
			push!(items[end], strip(line))
			last_was_blank = false
		end
	end
	items
end

day7() = eachline(joinpath(DATA, "day7.txt"))

day8() = eachline(joinpath(DATA, "day8.txt"))

day9() = parse.(Int, eachline(joinpath(DATA, "day9.txt")))

day10() = parse.(Int, eachline(joinpath(DATA, "day10.txt")))

function day11() 
	lines = eachline(joinpath(DATA, "day11.txt"))
	hcat(map(collect, lines)...)
end

day12() = eachline(joinpath(DATA, "day12.txt"))

function day13()
	lines = readlines(joinpath(DATA, "day13.txt"))
	(parse(Int, lines[1]), lines[2])
end

day14() = eachline(joinpath(DATA, "day14.txt"))

day16() = split(open(f -> read(f, String), joinpath(DATA, "day16.txt")), "\n\n")

end # module

File: 2020/fsharp/AOC/Parsers.fs
namespace AOC

open System

module Parsers = 
    let DATA = System.IO.Path.Combine [| __SOURCE_DIRECTORY__; ".."; ".."; "data" |]

    let day1 () =
        System.IO.Path.Combine [| DATA; "day1.txt" |]
        |> System.IO.File.ReadLines 
        |> Seq.map (fun x -> int(x))
        |> List.ofSeq

    let day2 () =
        System.IO.Path.Combine [| DATA; "day2.txt" |]
        |> System.IO.File.ReadLines 

    let day3() = 
        System.IO.Path.Combine [| DATA; "day3.txt" |]
        |> System.IO.File.ReadLines
        |> Seq.map (fun line -> Seq.map (fun x -> if x = '#' then 1 else 0) line)
        |> array2D

    let day4() = 
        let file = System.IO.Path.Combine [| DATA; "day4.txt" |]
                |> System.IO.File.ReadAllText 
        file.Split([|"\n\n"|], StringSplitOptions.RemoveEmptyEntries)

    let day5() =
        System.IO.Path.Combine [| DATA; "day5.txt" |]
        |> System.IO.File.ReadLines 

    let day6() = 
        let file = System.IO.Path.Combine [| DATA; "day6.txt" |]
                |> System.IO.File.ReadAllText
        file.Split([|"\n\n"|], StringSplitOptions.RemoveEmptyEntries)
        |> Seq.map (fun group -> group.Split([|'\n'|]))

    let day7() = 
        System.IO.Path.Combine [| DATA; "day7.txt" |]
        |> System.IO.File.ReadLines         

    let day8() = 
        System.IO.Path.Combine [| DATA; "day8.txt" |]
        |> System.IO.File.ReadLines  

    let day9() = 
        System.IO.Path.Combine [| DATA; "day9.txt" |]
        |> System.IO.File.ReadLines 
        |> Seq.map int64
        |> Array.ofSeq

    let day10() = 
        System.IO.Path.Combine [| DATA; "day10.txt" |]
        |> System.IO.File.ReadLines 
        |> Seq.map int
        |> Array.ofSeq    

    let day11() = 
        System.IO.Path.Combine [| DATA; "day11.txt" |]
        |> System.IO.File.ReadLines
        |> array2D

    let day12() = 
        System.IO.Path.Combine [| DATA; "day12.txt" |]
        |> System.IO.File.ReadLines    

    let day13() = 
        let file = System.IO.Path.Combine [| DATA; "day13.txt"|]
                    |> System.IO.File.ReadAllText
        let lines = file.Split([|"\n"|], StringSplitOptions.RemoveEmptyEntries)
        (int lines.[0], lines.[1])

    let day14() = 
        System.IO.Path.Combine [| DATA; "day14.txt" |]
        |> System.IO.File.ReadLines 

    let day16() = 
        let file = System.IO.Path.Combine [| DATA; "day16.txt" |]
                    |> System.IO.File.ReadAllText 
        file.Split([|"\n\n"|], StringSplitOptions.RemoveEmptyEntries)

    let day19General path = 
        let file = System.IO.Path.Combine [| DATA; path |]
                    |> System.IO.File.ReadAllText 
        let splat = file.Split([|"\n\n"|], StringSplitOptions.RemoveEmptyEntries)
        let rules = splat.[0].Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        let messages = splat.[1].Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        rules, messages


    let day19() = 
        day19General "day19.txt"
    
    let day19Test() = 
        day19General "day19_test.txt"
